### PR TITLE
monitoring: Allow arbitrary number of alerts per Observable

### PR DESCRIPTION
**Pull request by tsenart**
This commit is the first in a series of changes that address recent
pains with our monitoring generator. It refactors Observable to
have a slice of Alerts, rather than explicit Warning and Critical
fields, which allows us to have zero, or any number of alerts per panel.

### CODE REVIEW

**Comment by tsenart** [comment]
Most changes are very mechanical refactorings. Focus your review effort in `generator.go`.

**Comment by tsenart** [comment]
After this PR is merged, I want to change the PanelOptions to allow us to much more directly use the Grafana SDK.

File: monitoring/generator.go [hunk 1]
```diff
@@ -222,78 +217,71 @@ func (o Observable) validate() error {
...
+		} else if o.PossibleSolutions != "none" {
+			if _, err := goMarkdown(o.PossibleSolutions); err != nil {
+				return fmt.Errorf(`%s: PossibleSolutions: %v`, o.Name, err)
+			}
 		}
 	}
+
 	if o.Owner == "" {
-		return errors.New("Observable.Owner must be defined")
+		return fmt.Errorf("%s: has now Owner set", o.Name)
```
Comment (uwedeportivo) [inline monitoring/generator.go:246]:
```suggestion
		return fmt.Errorf("%s: has no Owner set", o.Name)
```

**Review by uwedeportivo — APPROVED** [review]
LGTM with the caveat that i'm not deeply immersed in the philosophy of alerting anymore (ie. the 5 pillars), so i don't know how this fits there and where the discussion stands on that.

**Review by bobheadxi — CHANGES_REQUESTED** [review]
> which allows us to have zero, or any number of alerts per panel.

Since the former can be possible with the explicit `Warning` and `Critical` fields (just remove the check for them), is the intent to set multiple `Warning` alerts? I'm concerned that the latter will look pretty bad in the panels (we generate "threshold" lines to indicate when a metric is above its alerting level, and even more of these beyond the warning/critical will get rather crowded), and also not make much sense - if you want more warnings, the right thing to do maybe is to introduce a new level of alerts. What is the path forward you envision with this change?

**Comment by tsenart** [comment]
> I'm concerned that the latter will look pretty bad in the panels (we generate "threshold" lines to indicate when a metric is above its alerting level

Ah, good point! I didn't think of that.

> if you want more warnings, the right thing to do maybe is to introduce a new level of alerts. What is the path forward and use case you envision with this change?

It seemed natural to me that you could have any number of alerts defined on a metric, but I didn't consider the constraint you mentioned above, so I'm happy to close this PR in favour of just removing the validation. Thanks!
.

**Pull request closed without merging**